# ML-03 — Frame Your Lane as an ML Task

This skeleton is yours to fill. Work the sections **in order** — each one has a one-line hint. Simple words, honest numbers.

> Working with an AI assistant? Tell it to read `skills/README.md` first and load the one skill this assignment names on its card.

## 1. My lane as an ML task (type)

*Classification, clustering, ranking, or scoring — which one, and why?*

The lane that I'm choosing is ranking/scoring.

My question is "which pages a revieweer will check first" not "what type of pages exist or if they are good or bad". Because I want an ordered list of pages to review, ranking/scoring fits better than classification or clustering.

Not classification: Because I think a ranked list is more useful anyway. Also beacuse I think that a    reviewer only has time to check few pages at once. Classification would mean putting each page into 
a fixed bucket, like "needs fix" or "fine." Also I dont have a clean yes/no answer for that.

Not clustering: Because clustering will group pages into types. I don't care what type a page is, why is the page falling short of expected clicks, that is what i care about, so scoring or ranking fits better.

So scoring each page by how far its CTR falls below similar pages, then 
sorting that list, is the closest fit to what I actually need.

## 2. Target or proxy

*What would you predict? Where does that label come from — observed outcome or a defined rule?*

The fact I want to predict/score is that, at the same position tier, how far a page's CTR will actually fall, below the median CTR of other pages. This can be called as "CTR gap".

This isn't a rule but something that i observed in the data(real CTR, real position tier). This matters because if i used something like "trend diection", it already comes from someone else's rule, so my model would just be learning that rule, not learning anything real about the data.

A bigger CTR gap means the page has a higher priority to check.

## 3. Success metric

*One metric you can defend. What number means 'good'?*

The metric that I'll ideally use will be "precision@K" - meaning, out of the 
top K pages I flag (like top 18), how many pages will actually need fixing when a real person checks them.

Right now I can't calculate this, because no one has manually 
reviewed these pages yet to say which ones "really" needed a fix. So 
for this week, what I can actually check is whether the CTR gap actually 
separates the pages in a meaningful way - like, is there a real difference 
between the top-ranked pages and the bottom-ranked pages, or is it mostly 
noise?

Once real reviewer feedback exists, precision@K becomes the honest way 
to see if my ranking is actually useful.

## 4. The unit of analysis, as a real dataframe

*Load your lane's slice and show it: one row = one what?*

In [2]:
import pandas as pd

df = pd.read_csv('../../data/raw/content_refresh_anonymized.csv')
visible = df[(df['impressions_90d'] >= 500) & (df['avg_position'] > 0) & (df['avg_position'] <= 20)].copy()

print(f"One row = one page. Total visible pages: {len(visible):,}")
visible[['content_id', 'client_id', 'position_tier', 'avg_position', 'ctr', 'impressions_90d']].head()

One row = one page. Total visible pages: 12,023


,content_id,client_id,position_tier,avg_position,ctr,impressions_90d
0,content_304f48230142,client_f369cb89fc,striking,10.6,0.76,3803
3,content_331d6c4de07b,client_19581e27de,page_1,6.2,0.49,11751
5,content_d4084a4bc775,client_f369cb89fc,page_1,8.5,0.03,3970
9,content_c27558df2b0c,client_19581e27de,page_1,4.9,0.16,1240
10,content_d8ee6cc6d642,client_19581e27de,top_3,2.2,1.55,20919


## 5. Why ML beats a fixed rule here

*What makes the pattern too messy for an if-statement?*

A simple rule like "flag pages with CTR below 0.1" won't work well here. 
That's because 0.1 CTR is actually bad if the page is at position 3, 
but totally fine if it's at position 15. So the "right" cutoff isn't 
one number, it changes depending on where the page ranks.

I could try writing a separate rule for each position tier, but even 
within one tier there's a lot of spread. Like last week, page_1 pages 
ranged all the way from 0 CTR to 5.42 CTR. That's too messy to handle 
with a few if-statements by hand across 12,023 pages. But code can 
group pages by tier and compare each one fairly, automatically, no 
matter how many pages there are.

## Self-check

Before you submit, confirm each line honestly:

- [x] Every section above is filled — markdown thinking AND the code that backs it
- [x] The notebook runs top to bottom with no errors (Runtime → Run all)
- [x] No client names, URLs, or private queries anywhere
- [x] My claims use careful words: observed, measured, directional, decision-support
- [ ] Committed to my repo under `work/notebooks/` — then submit your repo URL on the card. Done.